File: 2024-2 (IV Semestre)/Complejidad_Algorítmica/TallerGrupal3Complejidad/main.py
import random
import time


def selection_sort(arr):
    comparisons = 0
    swaps = 0
    n = len(arr)

    start_time = time.time_ns()

    for i in range(n):
        min_idx = i
        for j in range(i+1, n):
            comparisons += 1
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            swaps += 1
            arr[i], arr[min_idx] = arr[min_idx], arr[i]

    end_time = time.time_ns()  # tiempo en ns
    duration = end_time - start_time

    print(f"Stable: No")
    print(f"Comparisons: {comparisons}")
    print(f"Swaps: {swaps}")
    print(f"Insertions: No")
    print(f"Run Time: {duration} ns")
    print("Time Complexity: O(n^2)")


if __name__ == "__main__":
    n = int(input("Array Size:"))
    arr = [random.randint(300000, 302500) for _ in range(n)]
    selection_sort(arr)

File: 2024-2 (IV Semestre)/Complejidad_Algorítmica/TallerGrupal3Complejidad/main.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func selectionSort(arr []int) (comparisons int, swaps int, duration int64) {
	start := time.Now()

	for i := 0; i < len(arr)-1; i++ {
		minIndex := i
		for j := i + 1; j < len(arr); j++ {
			comparisons++
			if arr[j] < arr[minIndex] {
				minIndex = j
			}
		}
		if minIndex != i {
			swaps++
			arr[i], arr[minIndex] = arr[minIndex], arr[i]
		}
	}

	duration = time.Since(start).Nanoseconds()
	return
}

func generateRandomArray(size, min, max int) []int {
	arr := make([]int, size)
	for i := range arr {
		arr[i] = rand.Intn(max-min+1) + min
	}
	return arr
}

func main() {
	var n int
	fmt.Print("Array Size:")
	fmt.Scan(&n)
	arr := generateRandomArray(n, 300000, 302500)
	comparisons, swaps, duration := selectionSort(arr)

	fmt.Println("Stable: No")
	fmt.Printf("Comparisons: %d\n", comparisons)
	fmt.Printf("Swaps: %d\n", swaps)
	fmt.Println("Insertions: No")
	fmt.Printf("Run Time: %d ns\n", duration)
	fmt.Println("Time Complexity: O(n^2)")
}
